src/main/java/org/truffleruby/core/array/ArrayBuilderNode.java

```diff
@@ -69,475 +41,229 @@ public static ArrayBuilderNode create() {
+            final ArrayMirror mirror = arrayStrategy.newMirrorFromStore(array);
+            final ArrayMirror otherMirror = otherStrategy.newMirror(other);
+            final int otherSize = Layouts.ARRAY.getSize(other);
+            final int neededSize = index + otherSize;
+            final ArrayMirror newMirror;
+
+            if (neededSize > mirror.getLength()) {
+                replaceNodes(generalized, neededSize);
+                final int capacity = Math.max(mirror.getLength(), neededSize);
+                newMirror = generalized.newArray(capacity);
             } else {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                return appendValueFallback(store, index, value, expectedLength);
+                replaceNodes(generalized, mirror.getLength());
+                newMirror = generalized.newArray(mirror.getLength());
             }
-        }
 
-        @Override
-        public Object finish(Object store, int length) {
-            return store;
+            mirror.copyTo(newMirror, 0, 0, index);
+            otherMirror.copyTo(newMirror, 0, index, otherSize);
+            return newMirror.getArray();
         }
 
     }
 
-    private static class ObjectArrayBuilderNode extends ArrayBuilderNode {
-
-        private final int expectedLength;
-
-        @CompilationFinal private boolean hasAppendedObjectArray = false;
-        @CompilationFinal private boolean hasAppendedIntArray = false;
-        @CompilationFinal private boolean hasAppendedLongArray = false;
-        @CompilationFinal private boolean hasAppendedDoubleArray = false;
-
-        public ObjectArrayBuilderNode(int expectedLength) {
-            this.expectedLength = expectedLength;
-        }
-
-        @Override
-        public Object start() {
-            return new Object[expectedLength];
-        }
-
-        @Override
-        public Object start(int length) {
-            if (length > expectedLength) {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                final ArrayBuilderNode newNode = new ObjectArrayBuilderNode(length);
-                replace(newNode, length + " > " + expectedLength + " (expected)");
-                return newNode.start(length);
-            }
-
-            return new Object[expectedLength];
-        }
-
-        @Override
-        public Object ensure(Object store, int length) {
-            if (length > ((Object[]) store).length) {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                final ArrayBuilderNode newNode = new ObjectArrayBuilderNode(length);
-                replace(newNode, length + " > " + expectedLength + " (expected)");
-                return newNode.ensure(ArrayUtils.copyOf((Object[]) store, length), length);
-            }
+    @ImportStatic(ArrayGuards.class)
+    public abstract static class EnsureCapacityNode extends ArrayBuilderBaseNode {
```

Comment:
Doesn't need to be in this PR, but I think we can get rid of ArrayBuilderNode.ensure() since appendValue/appendArray already handle when they need to grow and the final length is learned by expectedLength in StartNode.
It's also only used in ArrayConcatNode.
Actually, since ArrayConcatNode is also the only user of appendArray, maybe we should have appendValue.
It might even be beneficial if for some reason an array is a Object[] but only contains primitives in ArrayConcatNode.